launcher/src/main/java/org/apache/spark/launcher/AbstractLauncher.java

```diff
@@ -0,0 +1,600 @@
+
+  protected String javaHome;
+  protected String sparkHome;
+  protected String propertiesFile;
+  protected final Map<String, String> conf = new HashMap<String, String>();
+  private final Map<String, String> env;
+
+  protected AbstractLauncher() {
+    this(null);
+  }
+
+  protected AbstractLauncher(Map<String, String> env) {
+    this.env = env;
+  }
+
+  @SuppressWarnings("unchecked")
+  private final T THIS = (T) this;
+
+  /** Set a custom JAVA_HOME for launching the Spark application. */
+  public T setJavaHome(String path) {
+    checkNotNull(path, "path");
+    this.javaHome = path;
+    return THIS;
+  }
+
+  /** Set a custom Spark installation location for the application. */
+  public T setSparkHome(String path) {
+    checkNotNull(path, "path");
+    this.sparkHome = path;
+    return THIS;
+  }
+
+  /** Set a custom properties file with Spark configuration for the application. */
+  public T setPropertiesFile(String path) {
+    checkNotNull(path, "path");
+    this.propertiesFile = path;
+    return THIS;
+  }
+
+  /** Set a single configuration value for the application. */
+  public T setConf(String key, String value) {
+    checkNotNull(key, "key");
+    checkNotNull(value, "value");
+    checkArgument(key.startsWith("spark."), "'key' must start with 'spark.'");
+    conf.put(key, value);
+    return THIS;
+  }
+
+  /**
+   * Launchers should implement this to create the command to be executed.
+   */
+  protected abstract List<String> buildLauncherCommand() throws IOException;
+
+  protected Properties loadPropertiesFile() throws IOException {
+    Properties props = new Properties();
+    File propsFile;
+    if (propertiesFile != null) {
+      propsFile = new File(propertiesFile);
+      checkArgument(propsFile.isFile(), "Invalid properties file '%s'.", propertiesFile);
+    } else {
+      String confDir = getenv("SPARK_CONF_DIR");
+      if (confDir == null) {
+        confDir = join(File.separator, getSparkHome(), "conf");
+      }
+      propsFile = new File(confDir, "spark-defaults.conf");
+    }
+
+    if (propsFile.isFile()) {
+      FileInputStream fd = null;
+      try {
+        fd = new FileInputStream(propsFile);
+        props.load(new InputStreamReader(fd, "UTF-8"));
+      } finally {
+        if (fd != null) {
+          try {
+            fd.close();
+          } catch (IOException e) {
+            // Ignore.
+          }
+        }
+      }
+    }
+
+    return props;
+  }
+
+  protected String getSparkHome() {
+    String path = first(sparkHome, getenv("SPARK_HOME"));
+    checkState(path != null,
+        "Spark home not found; set it explicitly or use the SPARK_HOME environment variable.");
+    return path;
+  }
+
+  protected List<String> createJavaCommand() throws IOException {
+    List<String> cmd = new ArrayList<String>();
+    if (javaHome == null) {
+      cmd.add(join(File.separator, System.getProperty("java.home"), "..", "bin", "java"));
```

Comment:
`java.home` generally points to the `jre` subdirectory. Probably works fine that way too.